Discovery Docs (#111)
LiqoTech/liqo: docs/design/discovery-component.md

# Discovery
## Overview
This component's goal is to find other clusters running Liqo around us, get information needed to pair and start peering process

### Features
List of supported features
* ClusterID creation
  * if not already present, during component starting, it creates new ClusterID taking the UID of first master of our cluster
* Make our cluster discoverable by other clusters

>   * this feature can be enabled and disabled runtime setting `enableAdvertisement` flag in `ClusterConfig` CR

Reviewer: ```suggestion
  * this feature can be enabled and disabled at runtime setting `enableAdvertisement` flag in `ClusterConfig` CR
```

  * register Liqo service on local mDNS server and answers when someone is looking for it
* Discovery on LAN

>   * this feature can be enabled and disabled runtime setting `enableDiscovery` flag in `ClusterConfig` CR

Reviewer: ```suggestion
  * this feature can be enabled and disabled at runtime setting `enableDiscovery` flag in `ClusterConfig` CR
```

>   * throw mDNS is able to find registered Liqo services on current LAN, a new `ForeignCluster` CR will be created for each cluster

Reviewer: what does "throw" mean?

* Discovery on WAN
  * when a new `SearchDomain` CR is added, an operator retrieves data from DNS server, a new `ForeignCluster` CR will be created for each cluster registered to domain provided
* Automatically join discovered clusters

>   * this feature can be enabled and disabled runtime setting `autojoin` flag in `ClusterConfig` CR

Reviewer: ```suggestion
  * this feature can be enabled and disabled at runtime setting `autojoin` flag in `ClusterConfig` CR
```

  * when new `ForeignCluster` is added, peering process will automatically begin
  * this value can be overwritten for WAN discovery setting it in `SearchDomain` CR
* Start peering process

>   * when `join` flag become true in a specific `ForeignCluster`, peering process will start for that cluster

Reviewer: ```suggestion
  * when `join` flag becomes true in a specific `ForeignCluster`, peering process will start for that cluster
```

### Limitations
List of known limitations
* After a `ForeignCluster` is created it will never disappear
* There is no way to set default behavior if to connect or not
  * as for Wi-Fi connections we may remember if automatically connect to a specific service when found
* CA used on the remote cluster has to be the same used by Deployments to authenticate API Server
  * no way to use URL as API server address (only raw IP is supported)
* Local cluster does not handle remote cluster CA changes
* It is not possible to trust or not CAs and to authenticate remote cluster

> * If API Server IP change, `ForeignCluster` will point to wrong one and will never change, leading to impossibility to contact remote cluster

Reviewer: ```suggestion
* If API Server IP changes, `ForeignCluster` will continue to point to the old one and will never change, leading to impossibility to contact remote cluster
```

* It needs a master in the cluster

## Architecture and workflow

This component can be divided in two main blocks:

1. Components in charge to find clusters and create `ForeignCluster`s
    * mDNS resolver
    * DNS client
    * Search Domain Operator
2. `ForeignCluster` management
    * Foreign Cluster Operator

#### Cluster finding

The goal of this block is to find clusters, to collect data and to create `ForeignCluster`s CR.

This can be done by mDNS resolver, DNS client or by manual insertion.

If we are using DNS client, we use an additional sub-component, the `SearchDomain` operator. This operator watches `SearchDomain` resources, when a new one is added, it contacts DNS server to retrieve required data.

##### Merge Logic

This logic merges discovered clusters with already existent ones. Currently, it checks if there is a cluster with the same ClusterId, if not it creates a new one.

#### ForeignCluster management

This block is a standard Kubernetes operator that is watching on `ForeignCluster` resources.

> When new one is added this component retrieves CAData from the remote cluster and stores it in a secret. That secret will be used in all next interactions with remote cluster to authenticate it.

Reviewer: this secret allows authentication of home cluster to foreign cluster or the opposite (authenticate foreign cluster for home cluster to know if he truly is what he says to be?) 
From the text I understand the second one

Author: yes, it's that. I tried to better specify now

> When `join` flag become true in a `ForeignCluster`, this component creates a new `PeeringRequest` CR in the remote cluster triggering peering process.

Reviewer: ```suggestion
When `join` flag becomes true in a `ForeignCluster`, this component creates a new `PeeringRequest` CR in the remote cluster triggering peering process.
```
